graylog2-server/src/main/java/org/graylog2/decorators/DecoratorProcessor.java

```diff
@@ -0,0 +1,30 @@
+/**
+ * This file is part of Graylog.
+ *
+ * Graylog is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Graylog is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Graylog.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.graylog2.decorators;
+
+import com.google.inject.ImplementedBy;
+import org.graylog2.indexer.results.ResultMessage;
+import org.graylog2.rest.resources.search.responses.SearchResponse;
+
+import java.util.List;
+import java.util.Optional;
+
+@ImplementedBy(DecoratorProcessorImpl.class)
+public interface DecoratorProcessor {
+    List<ResultMessage> decorate(List<ResultMessage> messages, Optional<String> stream);
+    SearchResponse decorate(SearchResponse searchResponse, Optional<String> stream);
```

Comment:
I'm not sure whether we really want both of these methods.
Currently, if a decorator creates a new field or additional messages, the metadata in `SearchResponse` is incorrect because of how it is assembled.
Additionally it seems like `decorate(SearchResponse)` includes what `decorate(List<ResultMessage>)` does. Both have access to the list of messages.

I'd vote for simplifying the interface to only process `SearchResponse`. I assume the first method is there to make it easier to implement a decorator, but I think then it becomes much harder to automatically keep the search metadata in sync (because we need to track the field names, result counts etc).

Additionally I think the result needs to be somehow annotated to allow the UI to hide the "search for field content" button: for synthetic fields there won't be anything to search for.